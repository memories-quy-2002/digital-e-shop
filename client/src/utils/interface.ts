export enum Role {
	Customer = "Customer",
	Admin = "Admin",
}

export interface User {
	UID: number;
	UName: string;
	UEmail: string;
	UPassword: string;
	URole: Role;
	ULast_login: string;
}
