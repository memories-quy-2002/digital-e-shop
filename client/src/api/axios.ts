import axios from "axios";
declare module "axios" {
    export interface AxiosRequestConfig {
        handlerEnabled?: boolean;
    }
}

const baseURL =
    process.env.NODE_ENV === "production"
        ? "https://e-commerce-express-server-app.vercel.app/"
        : "http://localhost:4000";

export default axios.create({
    baseURL,
    withCredentials: true,
    handlerEnabled: true,
});
