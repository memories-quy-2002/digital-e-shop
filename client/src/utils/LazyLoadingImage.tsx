import { useEffect, useRef } from "react";

const LazyLoadImage = ({ src, alt }: { src: string; alt: string }) => {
    const imgRef = useRef<HTMLImageElement | null>(null);

    useEffect(() => {
        const observer = new IntersectionObserver(
            (entries) => {
                entries.forEach((entry) => {
                    if (entry.isIntersecting && imgRef.current) {
                        imgRef.current.src = imgRef.current.dataset.src!;
                        observer.unobserve(imgRef.current);
                    }
                });
            },
            { threshold: 0.05 } // Load when 10% of the image is visible
        );

        if (imgRef.current) observer.observe(imgRef.current);
    }, []);

    return <img ref={imgRef} data-src={src} alt={alt} loading="lazy" />;
};
export default LazyLoadImage;
