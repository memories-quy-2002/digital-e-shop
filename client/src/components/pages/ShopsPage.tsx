import { useEffect, useState } from "react";
import { useLocation, useNavigate } from "react-router-dom";
import Cookies from "universal-cookie";
import axios from "../../api/axios";
import "../../styles/ShopsPage.scss";
import { Product } from "../../utils/interface";
import AsideShops from "../common/AsideShops";
import NavigationBar from "../common/NavigationBar";
import PaginatedItems from "../common/PaginatedItems";
import Layout from "../layout/Layout";

const cookies = new Cookies();
const ITEMS_PER_PAGE = 6;

type Wishlist = {
    id: number;
    product: Product;
};

type Filters = {
    term: string;
    categories: string[];
    brands: string[];
    priceRange: [number, number];
};

const ShopsPage = () => {
    const [products, setProducts] = useState<Product[]>([]);
    const [filteredProducts, setFilteredProducts] = useState<Product[]>([]);
    const [isLoading, setIsLoading] = useState<boolean>(false);
    const [wishlist, setWishlist] = useState<Wishlist[]>([]);
    const [filters, setFilters] = useState<Filters>({
        term: "",
        categories: [],
        brands: [],
        priceRange: [0, 4000],
    });

    const navigate = useNavigate();
    const location = useLocation();
    const uid =
        cookies.get("rememberMe")?.uid ||
        (sessionStorage["rememberMe"]
            ? JSON.parse(sessionStorage["rememberMe"]).uid
            : "");

    const updateURL = (newFilters: Filters) => {
        const queryParams = new URLSearchParams({
            categories: newFilters.categories.join(","),
            brands: newFilters.brands.join(","),
            minPrice: newFilters.priceRange[0].toString(),
            maxPrice: newFilters.priceRange[1].toString(),
            term: newFilters.term,
        });
        navigate(`${location.pathname}?${queryParams.toString()}`);
    };

    const handleTermChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        setFilters((prevFilters) => {
            return {
                ...prevFilters,
                term: e.target.value.trim(),
            };
        });
    };

    const handleCheckboxChange = (
        type: "categories" | "brands",
        value: string
    ) => {
        setFilters((prevFilters) => {
            const updatedArray = prevFilters[type].includes(value)
                ? prevFilters[type].filter((item) => item !== value) // Uncheck checkbox
                : [...prevFilters[type], value]; // Check checkbox

            return {
                ...prevFilters,
                [type]: updatedArray,
            };
        });
    };

    const handlePriceRangeChange = (newValue: [number, number]) => {
        setFilters((prevFilters) => ({
            ...prevFilters,
            priceRange: newValue,
        }));
    };

    const getFilteredProducts = (
        filters: Filters,
        allProducts: Product[]
    ): Product[] => {
        const { term, categories, brands, priceRange } = filters;

        return allProducts.filter((product) => {
            const matchCategory =
                categories.length === 0 ||
                categories.includes(product.category);

            const matchBrand =
                brands.length === 0 || brands.includes(product.brand);

            const matchPrice =
                product.price >= priceRange[0] &&
                product.price <= priceRange[1];

            const matchTerm =
                term.trim().toLowerCase() === ""
                    ? true
                    : product.name
                          .toLowerCase()
                          .includes(term.trim().toLowerCase());

            // Trả về sản phẩm nếu thoả mãn tất cả các điều kiện
            return matchCategory && matchBrand && matchPrice && matchTerm;
        });
    };

    const applyFilters = () => {
        updateURL(filters);
        setFilteredProducts(getFilteredProducts(filters, products));
    };

    useEffect(() => {
        const queryParams = new URLSearchParams(location.search);
        const newFilters: Filters = {
            term: queryParams.get("term") ?? "",
            categories:
                queryParams
                    .get("categories")
                    ?.split(",")
                    .filter((category) => category !== "") ?? [],
            brands:
                queryParams
                    .get("brands")
                    ?.split(",")
                    .filter((brand) => brand !== "") ?? [],
            priceRange: [
                Number(queryParams.get("minPrice") ?? 0),
                Number(queryParams.get("maxPrice") ?? 4000),
            ],
        };
        setFilters(newFilters);
        return () => {};
    }, [location.search]);

    useEffect(() => {
        const fetchProducts = async () => {
            setIsLoading(true); // Set loading state to indicate data fetching
            try {
                const response = await axios.get("/api/products");
                if (response.status === 200) {
                    setProducts(response.data.products);
                    setFilteredProducts(response.data.products);
                    console.log(response.data.msg);
                }
            } catch (err: any) {
                console.error(err);
            } finally {
                setIsLoading(false);
            }
        };
        fetchProducts();
        return () => {};
    }, []);

    useEffect(() => {
        const fetchWishlist = async () => {
            try {
                if (uid) {
                    const response = await axios.get(`/api/wishlist/${uid}`);
                    if (response.status === 200) {
                        const newWishlist: Wishlist[] =
                            response.data.wishlist.map((item: any) => {
                                const { id, product_id, ...productProps } =
                                    item;
                                return {
                                    id,
                                    product: {
                                        id: product_id,
                                        ...productProps,
                                    },
                                };
                            });

                        setWishlist(newWishlist);
                        console.log(response.data.msg);
                    }
                }
            } catch (err) {
                console.error(err);
            }
        };
        fetchWishlist();
    }, [uid]);

    return (
        <Layout>
            <NavigationBar />
            <div className="shops">
                <h2 style={{ marginBottom: "2rem", fontWeight: "bold" }}>
                    SHOPS PRODUCTS
                </h2>
                <div className="shops__container">
                    <AsideShops
                        products={products}
                        filters={filters}
                        onCheckboxChange={handleCheckboxChange}
                        onPriceRangeChange={handlePriceRangeChange}
                        onApplyFilters={applyFilters}
                        onTermChange={handleTermChange}
                    />
                    <div
                        data-testid="shops__container"
                        className="shops__container__main"
                    >
                        {isLoading && <p>Loading products...</p>}
                        <PaginatedItems
                            itemsPerPage={ITEMS_PER_PAGE}
                            items={filteredProducts}
                            uid={uid}
                            wishlist={wishlist}
                            isWishlistPage={false}
                        />
                    </div>
                </div>
            </div>
        </Layout>
    );
};
export default ShopsPage;
