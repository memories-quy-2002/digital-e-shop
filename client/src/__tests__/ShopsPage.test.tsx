import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import axios from "../api/axios"; // Import your custom Axios instance
import ShopsPage from "../components/pages/ShopsPage"; // Adjust based on your project structure
import ToastProvider from "../context/ToastContext";
import { Product } from "../utils/interface";

// Mock Axios
jest.mock("../api/axios");
const mockedAxios = axios as jest.Mocked<typeof axios>;

describe("ShopsPage", () => {
    const products: Product[] = [
        {
            id: 1,
            name: "Apple",
            category: "Electronics",
            brand: "Brand A",
            price: 99.99,
            sale_price: 79.99,
            rating: 4.5,
            reviews: 100,
            main_image: "https://example.com/product1.jpg",
            image_gallery: ["https://example.com/product1-1.jpg", "https://example.com/product1-2.jpg"],
            stock: 10,
            description: "This is a description of Product 1",
            specifications: ["Specification 1", "Specification 2"],
        },
        {
            id: 2,
            name: "Gucci",
            category: "Fashion",
            brand: "Brand B",
            price: 49.99,
            sale_price: null,
            rating: 4.2,
            reviews: 50,
            main_image: "https://example.com/product2.jpg",
            image_gallery: null,
            stock: 20,
            description: "This is a description of Product 2",
            specifications: ["Specification 3", "Specification 4"],
        },
        {
            id: 3,
            name: "LG",
            category: "Television",
            brand: "Brand C",
            price: 149.99,
            sale_price: null,
            rating: 4.8,
            reviews: 22,
            main_image: "https://example.com/product3.jpg",
            image_gallery: null,
            stock: 2,
            description: "This is a description of Product 3",
            specifications: ["Specification 5"],
        },
    ];

    afterEach(() => {
        jest.clearAllMocks();
    });

    it("matches the ShopsPage snapshot", async () => {
        const { asFragment } = render(
            <ToastProvider>
                <MemoryRouter>
                    <ShopsPage />
                </MemoryRouter>
            </ToastProvider>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders the shops page with products", async () => {
        mockedAxios.get.mockImplementationOnce(() => {
            return Promise.resolve({
                data: { products, msg: "Products are fetched successfully" },
                status: 200,
            });
        });
        render(
            <ToastProvider>
                <MemoryRouter>
                    <ShopsPage />
                </MemoryRouter>
            </ToastProvider>
        );

        expect(screen.getByText("SHOPS PRODUCTS")).toBeInTheDocument();
        await waitFor(() => {
            expect(mockedAxios.get).toHaveBeenCalledWith("/api/products");
        });
        await waitFor(() => {
            expect(screen.getByText("Apple")).toBeInTheDocument();
        });
        await waitFor(() => {
            expect(screen.getByText("Fashion")).toBeInTheDocument();
        });
    });

    it("renders the shop without products", async () => {
        render(
            <ToastProvider>
                <MemoryRouter>
                    <ShopsPage />
                </MemoryRouter>
            </ToastProvider>
        );

        await waitFor(() => {
            // Check for error message display
            expect(screen.getByText(/There is no product matched the filters/i)).toBeInTheDocument();
        });
    });

    it("renders the shops page with products filtered by search terms", async () => {
        mockedAxios.get.mockImplementationOnce(() => {
            return Promise.resolve({
                data: { products, msg: "Products are fetched successfully" },
                status: 200,
            });
        });
        render(
            <ToastProvider>
                <MemoryRouter>
                    <ShopsPage />
                </MemoryRouter>
            </ToastProvider>
        );

        await waitFor(() => {
            expect(screen.getByText("LG")).toBeInTheDocument();
        });
        const inputText = screen.getByPlaceholderText("Search product...");
        const applyButton = screen.getByRole("button", {
            name: /apply/i,
        });

        fireEvent.change(inputText, { target: { value: "LG" } });
        fireEvent.click(applyButton);
        await waitFor(() => {
            expect(screen.queryByText("Apple")).not.toBeInTheDocument();
        });
    });

    it("renders the shops page with no products matched filters by term", async () => {
        mockedAxios.get.mockImplementationOnce(() => {
            return Promise.resolve({
                data: { products, msg: "Products are fetched successfully" },
                status: 200,
            });
        });

        render(
            <ToastProvider>
                <MemoryRouter>
                    <ShopsPage />
                </MemoryRouter>
            </ToastProvider>
        );

        await waitFor(() => {
            expect(screen.getByText("LG")).toBeInTheDocument();
        });

        const inputText = screen.getByPlaceholderText("Search product...");
        const applyButton = screen.getByRole("button", {
            name: /apply/i,
        });

        fireEvent.change(inputText, { target: { value: "Not__filtered" } });
        fireEvent.click(applyButton);
        await waitFor(() => {
            expect(screen.queryByText("Apple")).not.toBeInTheDocument();
        });
        expect(screen.queryByRole("paragraph", { name: "Fashion" })).not.toBeInTheDocument();
        expect(screen.queryByText("LG")).not.toBeInTheDocument();
    });

    it("renders the shops page with products filtered by checkbox", async () => {
        mockedAxios.get.mockImplementationOnce(() => {
            return Promise.resolve({
                data: { products, msg: "Products are fetched successfully" },
                status: 200,
            });
        });

        render(
            <ToastProvider>
                <MemoryRouter>
                    <ShopsPage />
                </MemoryRouter>
            </ToastProvider>
        );
        await waitFor(() => {
            expect(screen.getByLabelText("Fashion")).toBeInTheDocument();
        });

        // const element = screen.getByTestId("shops__aside__brand");
        // console.log(prettyDOM(element));
        const checkbox = screen.getByLabelText("Fashion");
        const applyButton = screen.getByRole("button", {
            name: /apply/i,
        });

        expect(checkbox).toBeInTheDocument();
        expect(checkbox).not.toBeChecked();
        fireEvent.click(checkbox);
        expect(checkbox).toBeChecked();
        fireEvent.click(applyButton);

        await waitFor(() => {
            expect(screen.queryByText("Apple")).not.toBeInTheDocument();
        });
        expect(screen.getByText("Gucci")).toBeInTheDocument();
    });
});
