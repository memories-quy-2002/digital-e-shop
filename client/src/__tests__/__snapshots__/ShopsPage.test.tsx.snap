// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShopsPage matches the ShopsPage snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="header__container"
    >
      <div
        class="header__container__info"
      >
        <div
          class="header__container__info__personal"
        >
          <div
            class="header__container__info__personal__item"
          >
            <svg
              color="white"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              style="color: white;"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M391 480c-19.52 0-46.94-7.06-88-30-49.93-28-88.55-53.85-138.21-103.38C116.91 298.77 93.61 267.79 61 208.45c-36.84-67-30.56-102.12-23.54-117.13C45.82 73.38 58.16 62.65 74.11 52a176.3 176.3 0 0 1 28.64-15.2c1-.43 1.93-.84 2.76-1.21 4.95-2.23 12.45-5.6 21.95-2 6.34 2.38 12 7.25 20.86 16 18.17 17.92 43 57.83 52.16 77.43 6.15 13.21 10.22 21.93 10.23 31.71 0 11.45-5.76 20.28-12.75 29.81-1.31 1.79-2.61 3.5-3.87 5.16-7.61 10-9.28 12.89-8.18 18.05 2.23 10.37 18.86 41.24 46.19 68.51s57.31 42.85 67.72 45.07c5.38 1.15 8.33-.59 18.65-8.47 1.48-1.13 3-2.3 4.59-3.47 10.66-7.93 19.08-13.54 30.26-13.54h.06c9.73 0 18.06 4.22 31.86 11.18 18 9.08 59.11 33.59 77.14 51.78 8.77 8.84 13.66 14.48 16.05 20.81 3.6 9.53.21 17-2 22-.37.83-.78 1.74-1.21 2.75a176.49 176.49 0 0 1-15.29 28.58c-10.63 15.9-21.4 28.21-39.38 36.58A67.42 67.42 0 0 1 391 480z"
              />
            </svg>
            (+84) 123 456 7890
          </div>
          <div
            class="header__container__info__personal__item"
          >
            <svg
              color="white"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              style="color: white;"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M464 80H48a16 16 0 0 0-16 16v320a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V96a16 16 0 0 0-16-16zM265.82 284.63a16 16 0 0 1-19.64 0L89.55 162.81l19.64-25.26L256 251.73l146.81-114.18 19.64 25.26z"
              />
            </svg>
            digital-e@gmail.com
          </div>
          <div
            class="header__container__info__personal__item"
          >
            <svg
              color="white"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              style="color: white;"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M261.56 101.28a8 8 0 0 0-11.06 0L66.4 277.15a8 8 0 0 0-2.47 5.79L63.9 448a32 32 0 0 0 32 32H192a16 16 0 0 0 16-16V328a8 8 0 0 1 8-8h80a8 8 0 0 1 8 8v136a16 16 0 0 0 16 16h96.06a32 32 0 0 0 32-32V282.94a8 8 0 0 0-2.47-5.79z"
              />
              <path
                d="m490.91 244.15-74.8-71.56V64a16 16 0 0 0-16-16h-48a16 16 0 0 0-16 16v32l-57.92-55.38C272.77 35.14 264.71 32 256 32c-8.68 0-16.72 3.14-22.14 8.63l-212.7 203.5c-6.22 6-7 15.87-1.34 22.37A16 16 0 0 0 43 267.56L250.5 69.28a8 8 0 0 1 11.06 0l207.52 198.28a16 16 0 0 0 22.59-.44c6.14-6.36 5.63-16.86-.76-22.97z"
              />
            </svg>
            123 ABC Street, HCM City
          </div>
        </div>
        <div
          class="header__container__info__auth"
        >
          <strong>
            Welcome Anonymous
          </strong>
          <div
            class="header__container__info__auth__button"
          >
            <div>
              <a
                href="/login"
              >
                Login
              </a>
            </div>
            <div>
              |
            </div>
            <div>
              <a
                href="/signup"
              >
                Signup
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="header__container__main"
      >
        <div
          class="header__container__main__brand"
        >
          <a
            class="header__container__main__brand__link navbar-brand"
            href="/"
          >
            DIGITAL-E
          </a>
        </div>
        <div
          class="header__container__main__search"
        >
          <input
            class="header__container__main__search__bar"
            id=""
            name=""
            placeholder="Search"
            type="text"
          />
          <button>
            Search
          </button>
        </div>
        <div
          class="header__container__main__group"
        >
          <div
            class="header__container__main__group__item"
          >
            <svg
              fill="currentColor"
              height="28"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="28"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 448a32 32 0 0 1-18-5.57c-78.59-53.35-112.62-89.93-131.39-112.8-40-48.75-59.15-98.8-58.61-153C48.63 114.52 98.46 64 159.08 64c44.08 0 74.61 24.83 92.39 45.51a6 6 0 0 0 9.06 0C278.31 88.81 308.84 64 352.92 64c60.62 0 110.45 50.52 111.08 112.64.54 54.21-18.63 104.26-58.61 153-18.77 22.87-52.8 59.45-131.39 112.8a32 32 0 0 1-18 5.56z"
              />
            </svg>
            Wishlist
          </div>
          <div
            class="header__container__main__group__item"
          >
            <svg
              fill="currentColor"
              height="28"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="28"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="176"
                cy="416"
                r="32"
              />
              <circle
                cx="400"
                cy="416"
                r="32"
              />
              <path
                d="M456.8 120.78a23.92 23.92 0 0 0-18.56-8.78H133.89l-6.13-34.78A16 16 0 0 0 112 64H48a16 16 0 0 0 0 32h50.58l45.66 258.78A16 16 0 0 0 160 368h256a16 16 0 0 0 0-32H173.42l-5.64-32h241.66A24.07 24.07 0 0 0 433 284.71l28.8-144a24 24 0 0 0-5-19.93z"
              />
            </svg>
            Shopping Cart
          </div>
        </div>
      </div>
    </div>
    <nav
      class="navbar navbar-expand navbar-light bg-gray"
    >
      <div
        class="navbar navbar-nav"
      >
        <a
          class="navbar__item nav-link"
          data-rr-ui-event-key="/"
          href="/"
        >
          Home
        </a>
        <a
          class="navbar__item nav-link"
          data-rr-ui-event-key="/shops"
          href="/shops"
        >
          shops
        </a>
        <a
          class="navbar__item nav-link"
          data-rr-ui-event-key="/promotions"
          href="/promotions"
        >
          promotions
        </a>
        <a
          class="navbar__item nav-link"
          data-rr-ui-event-key="/about-us"
          href="/about-us"
        >
          about us
        </a>
        <a
          class="navbar__item nav-link"
          data-rr-ui-event-key="/contact-us"
          href="/contact-us"
        >
          contact us
        </a>
        <a
          class="navbar__item nav-link"
          data-rr-ui-event-key="/support"
          href="/support"
        >
          support
        </a>
        <a
          class="navbar__item nav-link"
          data-rr-ui-event-key="/news"
          href="/news"
        >
          news
        </a>
      </div>
    </nav>
    <div
      class="shops"
    >
      <h2
        style="margin-bottom: 2rem; font-weight: bold;"
      >
        SHOPS PRODUCTS
      </h2>
      <div
        class="shops__container"
      >
        <div
          class="shops__container__aside"
        >
          <div
            class="shops__container__aside__search"
          >
            <input
              placeholder="Search product..."
              type="text"
              value=""
            />
          </div>
          <div
            class="shops__container__aside__categories"
          >
            <div>
              <h4>
                Categories
              </h4>
            </div>
          </div>
          <div
            class="shops__container__aside__brands"
            data-testid="shops__aside__brand"
          >
            <div>
              <h4>
                Brands
              </h4>
            </div>
          </div>
          <div
            class="shops__container__aside__price"
          >
            <h4>
              Price range
            </h4>
            <div
              class="shops__container__aside__price__slider"
            >
              <div
                class="horizontal-slider"
                style="position: relative;"
              >
                <div
                  class="example-track example-track-0"
                  style="position: absolute; left: 0px; right: 0px;"
                />
                <div
                  class="example-track example-track-1"
                  style="position: absolute; left: 0px; right: 0px;"
                />
                <div
                  class="example-track example-track-2"
                  style="position: absolute; left: 0px; right: 0px;"
                />
                <div
                  aria-disabled="false"
                  aria-orientation="horizontal"
                  aria-valuemax="4000"
                  aria-valuemin="0"
                  aria-valuenow="0"
                  class="example-thumb example-thumb-0 "
                  role="slider"
                  style="position: absolute; z-index: 1; left: 0px;"
                  tabindex="0"
                />
                <div
                  aria-disabled="false"
                  aria-orientation="horizontal"
                  aria-valuemax="4000"
                  aria-valuemin="0"
                  aria-valuenow="4000"
                  class="example-thumb example-thumb-1 "
                  role="slider"
                  style="position: absolute; z-index: 2; left: 0px;"
                  tabindex="0"
                />
              </div>
              <div
                class="shops__container__aside__price__slider__num"
              >
                <span>
                  0
                </span>
                <span>
                  4000
                </span>
              </div>
            </div>
          </div>
          <button
            class="btn btn-info shops__container__aside__button"
            type="button"
          >
            Apply
          </button>
        </div>
        <div
          class="shops__container__main"
          data-testid="shops__container"
        >
          <p>
            Loading products...
          </p>
          <div
            class="shops__container__main__pagination"
          >
            <div
              style="display: flex; justify-content: center; align-items: center; height: 400px;"
            >
              <strong
                style="font-size: 1.5rem;"
              >
                There is no product matched the filters
              </strong>
            </div>
            <div
              style="display: flex; justify-content: center;"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="footer"
    >
      <div
        class="footer__menuGroup"
      >
        <div
          class="footer__menuGroup__menu"
        >
          <h5
            class="footer__menuGroup__menu__title"
          >
            Top Categories
          </h5>
          <div
            class="footer__menuGroup__menu__item"
          >
            <a
              href="/shops?categories=Laptops"
            >
              Laptops
            </a>
          </div>
          <div
            class="footer__menuGroup__menu__item"
          >
            <a
              href="/shops?categories=Cameras"
            >
              Cameras
            </a>
          </div>
          <div
            class="footer__menuGroup__menu__item"
          >
            <a
              href="/shops?categories=Smartphones"
            >
              Smartphones
            </a>
          </div>
          <div
            class="footer__menuGroup__menu__item"
          >
            <a
              href="/shops?categories=Graphic_Cards"
            >
              Graphic cards
            </a>
          </div>
        </div>
        <div
          class="footer__menuGroup__menu"
        >
          <h5
            class="footer__menuGroup__menu__title"
          >
            Help
          </h5>
          <div
            class="footer__menuGroup__menu__item"
          >
            Delivery
          </div>
          <div
            class="footer__menuGroup__menu__item"
          >
            Track an order
          </div>
          <div
            class="footer__menuGroup__menu__item"
          >
            Secure payments
          </div>
        </div>
        <div
          class="footer__menuGroup__menu"
        >
          <h5
            class="footer__menuGroup__menu__title"
          >
            About us
          </h5>
          <div
            class="footer__menuGroup__menu__item"
          >
            Careers
          </div>
          <div
            class="footer__menuGroup__menu__item"
          >
            Terms and Conditions
          </div>
          <div
            class="footer__menuGroup__menu__item"
          >
            Privacies and Policies
          </div>
        </div>
        <div
          class="footer__menuGroup__menu"
        >
          <h5
            class="footer__menuGroup__menu__title"
          >
            Feedback
          </h5>
          <div
            class="footer__menuGroup__menu__item"
          >
            Leave a feedback
          </div>
        </div>
        <div
          class="footer__menuGroup__menu"
        >
          <h5
            class="footer__menuGroup__menu__title"
          >
            Contact us
          </h5>
          <div
            class="footer__menuGroup__menu__item-disabled"
          >
            Email: digital-e@gmail.com
          </div>
          <div
            class="footer__menuGroup__menu__item-disabled"
          >
            Phone: (+84) 123 456 7890
          </div>
          <div
            class="footer__menuGroup__menu__item-disabled"
          >
            Address: 123 ABC Street, Ho Chi Minh City, Vietnam
          </div>
        </div>
      </div>
      <div
        class="footer__otherGroup"
      >
        <div
          class="footer__otherGroup__menu"
        >
          <h5
            class="footer__otherGroup__menu__title"
          >
            Find a store
          </h5>
          <div
            class="footer__otherGroup__menu__item-disabled"
          >
            Find Digital-E stores near you
          </div>
          <button
            class="btn"
            style="width: 250px;"
            type="button"
          >
            Store location
          </button>
        </div>
        <div
          class="footer__otherGroup__menu"
        >
          <h5>
            Follow Us
          </h5>
          <ul>
            <li>
              <a
                href="https://www.facebook.com/yourpage"
                rel="noopener noreferrer"
                target="_blank"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 320 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M279.14 288l14.22-92.66h-88.91v-60.13c0-25.35 12.42-50.06 52.24-50.06h40.42V6.26S260.43 0 225.36 0c-73.22 0-121.08 44.38-121.08 124.72v70.62H22.89V288h81.39v224h100.17V288z"
                  />
                </svg>
                 
                <span>
                  Facebook
                </span>
              </a>
            </li>
            <li>
              <a
                href="https://www.twitter.com/yourprofile"
                rel="noopener noreferrer"
                target="_blank"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 512 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"
                  />
                </svg>
                 
                <span>
                  Twitter
                </span>
              </a>
            </li>
            <li>
              <a
                href="https://www.instagram.com/yourprofile"
                rel="noopener noreferrer"
                target="_blank"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 448 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M224.1 141c-63.6 0-114.9 51.3-114.9 114.9s51.3 114.9 114.9 114.9S339 319.5 339 255.9 287.7 141 224.1 141zm0 189.6c-41.1 0-74.7-33.5-74.7-74.7s33.5-74.7 74.7-74.7 74.7 33.5 74.7 74.7-33.6 74.7-74.7 74.7zm146.4-194.3c0 14.9-12 26.8-26.8 26.8-14.9 0-26.8-12-26.8-26.8s12-26.8 26.8-26.8 26.8 12 26.8 26.8zm76.1 27.2c-1.7-35.9-9.9-67.7-36.2-93.9-26.2-26.2-58-34.4-93.9-36.2-37-2.1-147.9-2.1-184.9 0-35.8 1.7-67.6 9.9-93.9 36.1s-34.4 58-36.2 93.9c-2.1 37-2.1 147.9 0 184.9 1.7 35.9 9.9 67.7 36.2 93.9s58 34.4 93.9 36.2c37 2.1 147.9 2.1 184.9 0 35.9-1.7 67.7-9.9 93.9-36.2 26.2-26.2 34.4-58 36.2-93.9 2.1-37 2.1-147.8 0-184.8zM398.8 388c-7.8 19.6-22.9 34.7-42.6 42.6-29.5 11.7-99.5 9-132.1 9s-102.7 2.6-132.1-9c-19.6-7.8-34.7-22.9-42.6-42.6-11.7-29.5-9-99.5-9-132.1s-2.6-102.7 9-132.1c7.8-19.6 22.9-34.7 42.6-42.6 29.5-11.7 99.5-9 132.1-9s102.7-2.6 132.1 9c19.6 7.8 34.7 22.9 42.6 42.6 11.7 29.5 9 99.5 9 132.1s2.7 102.7-9 132.1z"
                  />
                </svg>
                 
                <span>
                  Instagram
                </span>
              </a>
            </li>
            <li>
              <a
                href="https://www.linkedin.com/in/yourprofile"
                rel="noopener noreferrer"
                target="_blank"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 448 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M100.28 448H7.4V148.9h92.88zM53.79 108.1C24.09 108.1 0 83.5 0 53.8a53.79 53.79 0 0 1 107.58 0c0 29.7-24.1 54.3-53.79 54.3zM447.9 448h-92.68V302.4c0-34.7-.7-79.2-48.29-79.2-48.29 0-55.69 37.7-55.69 76.7V448h-92.78V148.9h89.08v40.8h1.3c12.4-23.5 42.69-48.3 87.88-48.3 94 0 111.28 61.9 111.28 142.3V448z"
                  />
                </svg>
                 
                <span>
                  LinkedIn
                </span>
              </a>
            </li>
          </ul>
        </div>
        <div
          class="footer__otherGroup__menu"
        >
          <h5
            class="footer__otherGroup__menu__title"
          >
            Newsletter subscription
          </h5>
          <div
            class="footer__otherGroup__menu__item-disabled"
          >
            Receive product news and updates to your inbox
          </div>
          <div
            class="d-flex"
          >
            <input
              class="ps-2"
              id="email_subs"
              name="email_subs"
              placeholder="Your email"
              type="email"
            />
            <button
              class="btn"
              type="button"
            >
              Go
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="toast-container bottom-0 end-0 p-3"
    style="z-index: 1; position: fixed; bottom: 0px; right: 0px;"
  />
</DocumentFragment>
`;
